Implement Events Archiving on S3

akouthoofd opened the pull request:
Following the review of the file implementation I'm rethinking the design a bit. The currently implementation holds a reference to an ArchiverOnFile, but a better approach might be to extend off of that implementation instead, so we can keep the methods `protected` instead of `public`. Plus considering the current implementations reliance on the file archiver it may as well be an extension of. Thoughts?

akouthoofd committed:
Make archive idempotent

- events that have been restored in the past will not be archived a second time

akouthoofd committed:
Rebase and refactor

akouthoofd committed:
Implement Events archiving on S3 using ArchiverOnFile

akouthoofd committed:
Fix minor bug and minor refactor

- restore was pulling more files down than necessary

akouthoofd committed:
Update classes to adjust for method removed from interface

akouthoofd committed:
Fix expiration overriding already archived events

- minor bug fix for namespace check

akouthoofd committed:
Add additional test

akouthoofd committed:
Add basic file caching to reduce calls to S3

akouthoofd committed:
Batch store events during restore

akouthoofd committed:
Fix test

akouthoofd committed:
Minor clean up after rebase

akouthoofd committed:
Minor clean up after rebase

akouthoofd committed:
Fix bug

- ceiling should be on millis below not about rounding

akouthoofd committed:
Add optimization and minor refactor

- added extra comments

akouthoofd committed:
Prevent file uploads if they aren't necessary

akouthoofd committed:
Handle the case of archive files already in local storage on startup

pteyer commented on cantor-archive/src/main/java/com/salesforce/cantor/archive/file/ObjectsArchiverOnFile.java:
```diff
@@ -92,6 +92,7 @@ public void doRestore(final Objects objects, final String namespace, final Path
 
     public Path getFileArchive(final String namespace) {
-        return getFile(archivePathFormat, namespace);
+        return getFile(archivePathFormat,
```
one line.

akouthoofd replied:
Fixed.

pteyer commented on cantor-archive/src/main/java/com/salesforce/cantor/archive/file/SetsArchiverOnFile.java:
```diff
@@ -85,6 +85,7 @@ public void doRestore(final Sets sets, final String namespace, final Path archiv
 
     public Path getFileArchive(final String namespace) {
-        return getFile(archivePathFormat, namespace);
+        return getFile(archivePathFormat,
```
^^

akouthoofd replied:
Fixed.

pteyer commented on cantor-archive/src/main/java/com/salesforce/cantor/archive/s3/ArchiverOnS3.java:
```diff
@@ -1,49 +1,44 @@
-
-    protected static final String defaultBucketName = "cantor-archive";
-    protected static final long defaultChunkMillis = TimeUnit.HOURS.toMillis(1);
+    private static final String defaultArchivePathBase = "/tmp/cantor-s3-archive";
```
let's use cwd for this; so just `cantor-s3-archive` instead of `/tmp/...`

akouthoofd replied:
Yeah, I made this change in the next PR. I'll propagate it up.

akouthoofd committed:
Refactor S3 archiver to extend file archiver

akouthoofd committed:
Code clean up

- using pwd for archive location

akouthoofd committed:
Update tests

akouthoofd committed:
Remove extension of ArchiverOnFile

- extending that class was creating a second set of archiver instances that weren't used

akouthoofd committed:
Merge pull request #12 from AlexanderKouthoofd/alex/s3-archiver/extend-file

Refactor S3 archiver to extend file archiver

akouthoofd committed:
Merge pull request #11 from AlexanderKouthoofd/alex/s3-archiver/caching

Add optimizations and caching to Events Archiver on S3
